tools/lib/h5diff.c

```diff
@@ -1162,339 +1154,115 @@ diff_match(hid_t file1_id, const char *grp1, trav_info_t *info1, hid_t file2_id,
-                NULL) {
-                H5TOOLS_ERROR(H5DIFF_ERR, "name buffer allocation failed");
-            }
-            else {
-                strcpy(obj2_fullpath, grp2_path);
-                strcat(obj2_fullpath, table->objs[i].name);
-            }
-#endif /* H5_HAVE_ASPRINTF */
-                H5TOOLS_DEBUG("diff_match path2 - %s", obj2_fullpath);
-
-                /* get index to figure out type of the object in file1 */
-                while (info1->paths[idx1].path && (strcmp(obj1_fullpath, info1->paths[idx1].path) != 0))
-                    idx1++;
-                /* get index to figure out type of the object in file2 */
-                while (info2->paths[idx2].path && (strcmp(obj2_fullpath, info2->paths[idx2].path) != 0))
-                    idx2++;
-
-                /* Set argdata to pass other args into diff() */
-                argdata.type[0]        = info1->paths[idx1].type;
-                argdata.type[1]        = info2->paths[idx2].type;
-                argdata.is_same_trgobj = table->objs[i].is_same_trgobj;
-
-                opts->cmn_objs = 1;
-                if (!g_Parallel) {
-                    H5TOOLS_DEBUG("diff paths - errstat:%d", opts->err_stat);
-                    nfound += diff(file1_id, obj1_fullpath, file2_id, obj2_fullpath, opts, &argdata);
-                } /* end if */
-#ifdef H5_HAVE_PARALLEL
-                else {
-                    int workerFound = 0;
-
-                    H5TOOLS_DEBUG("Beginning of big else block");
-                    /* We're in parallel mode */
-                    /* Since the data type of diff value is hsize_t which can
-                     * be arbitrary large such that there is no MPI type that
-                     * matches it, the value is passed between processes as
-                     * an array of bytes in order to be portable.  But this
-                     * may not work in non-homogeneous MPI environments.
-                     */
-
-                    /*Set up args to pass to worker task. */
-                    if (strlen(obj1_fullpath) > 255 || strlen(obj2_fullpath) > 255) {
-                        fprintf(stderr,
-                                "The parallel diff only supports object names up to 255 characters\n");
-                        MPI_Abort(MPI_COMM_WORLD, 0);
-                    } /* end if */
-
-                    /* set args struct to pass */
-                    strcpy(args.name1, obj1_fullpath);
-                    strcpy(args.name2, obj2_fullpath);
-                    args.opts                   = *opts;
-                    args.argdata.type[0]        = info1->paths[idx1].type;
-                    args.argdata.type[1]        = info2->paths[idx2].type;
-                    args.argdata.is_same_trgobj = table->objs[i].is_same_trgobj;
-
-                    /* if there are any outstanding print requests, let's handle one. */
-                    if (busyTasks > 0) {
```

Comment:
Just about everything below this line has been refactored into the two `handle_worker_request` and `dispatch_diff_to_worker` worker routines and significantly condensed. `dispatch_diff_to_worker` is simply responsible for sending the `struct diff_mpi_args` struct to a worker task to tell it what operation to perform, while `handle_worker_request` is responsible for most of the refactored logic here that handles messages coming back from worker tasks.

This logic was essentially duplicated between here and the loop below with the comment `/* make sure all tasks are done */` that runs after all diff items have been handled, but wasn't quite identically duplicated and had several cases where the logic could end up in "invalid" states or be between output printing states and result in interleaved output. Now, the logic has been condensed to:

- looping over items to be diffed
  - handle incoming messages from a worker task if no task is available to dispatch work to
  - dispatch the diff work for an item to be diffed to a worker task
- while there are still busy worker tasks, handle incoming messages from a worker task until all are done

It's fairly similar to the previous logic, but without the previous in-between states the logic could get into between the main and secondary loops due to the way messages are handled between different MPI implementations.